File: java-provisioning/src/main/java/com/lastpass/provisioning/UsersProvisioning.java
package com.lastpass.provisioning;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.lastpass.common.crypto.RSACipher;
import com.lastpass.common.crypto.RSAUtils;
import com.lastpass.common.model.GenericRequest;
import com.lastpass.common.model.GenericResponse;
import com.lastpass.common.utils.GenericResponseHandler;
import com.lastpass.common.utils.RestClient;
import com.lastpass.provisioning.model.Group;
import com.lastpass.provisioning.model.Pagination;
import com.lastpass.provisioning.model.ProvisioningResponse;
import com.lastpass.provisioning.model.User;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.List;
import java.util.Map;
import org.apache.log4j.Logger;

/**
 * Consumes LastPass provisioning API services
 *
 * @author LogMeIn
 */
public class UsersProvisioning {

    private static final Logger LOG = Logger.getLogger(UsersProvisioning.class);

    private static String BASE_URL;
    private static String ADD_USER_URL;
    private static String EDIT_USER_URL;
    private static String DELETE_USER_URL;
    private static String SUSPEND_USER_URL;
    private static String RESEND_ACTIVATION_EMAIL_URL;
    private static String PUBLIC_KEY_REGISTRATION_URL;
    private static String IS_USER_EXISTS_URL;
    private static String ADD_GROUP_URL;
    private static String GET_GROUPS;
    private static String DELETE_GROUP_URL;
    private static String ASSIGN_GROUP_MEMBER_URL;
    private static String UNASSIGN_GROUP_MEMBER_URL;
    private static String POST_ROLE_URL;
    private static String GET_ROLES_URL;
    private static String DELETE_ROLE_URL;
    private static String ASSIGN_TO_ROLE_URL;
    private static String GET_ASSIGNED_TO_ROLE_URL;

    private final RSACipher rsaCipher;
    private final String genericAPIKey;

    /**
     *
     * @param rsaPublicKey
     * @param rsaPrivateKey
     * @param genericAPIKey
     * @param baseURL
     * @throws IOException
     */
    public UsersProvisioning(PublicKey rsaPublicKey, PrivateKey rsaPrivateKey, String genericAPIKey, String baseURL) throws IOException {
        this.genericAPIKey = genericAPIKey;
        this.rsaCipher = new RSACipher(rsaPublicKey, rsaPrivateKey, StandardCharsets.UTF_16LE);

        BASE_URL = baseURL;
        ADD_USER_URL = BASE_URL + "AddUser";
        EDIT_USER_URL = BASE_URL + "EditUser";
        DELETE_USER_URL = BASE_URL + "DeleteUser";
        SUSPEND_USER_URL = BASE_URL + "SuspendUser";
        RESEND_ACTIVATION_EMAIL_URL = BASE_URL + "ResendActivationEmail";
        PUBLIC_KEY_REGISTRATION_URL = BASE_URL + "PKReg";
        IS_USER_EXISTS_URL = BASE_URL + "IsUserExist";
        ADD_GROUP_URL = BASE_URL + "AddGroup";
        DELETE_GROUP_URL = BASE_URL + "DeleteGroup";
        GET_GROUPS = BASE_URL + "GetGroups";
        ASSIGN_GROUP_MEMBER_URL = BASE_URL + "AssignGroupMember";
        UNASSIGN_GROUP_MEMBER_URL = BASE_URL + "UnassignGroupMember";
        POST_ROLE_URL = BASE_URL + "PostRole";
        GET_ROLES_URL = BASE_URL + "GetRoles";
        DELETE_ROLE_URL = BASE_URL + "DeleteRole";
        ASSIGN_TO_ROLE_URL = BASE_URL + "AssignToRole";
        GET_ASSIGNED_TO_ROLE_URL = BASE_URL + "GetAssignedToRole";
    }

    /**
     * Creates a UsersProvisioning object
     *
     * @param publicKeyFile Path to RSA public key file
     * @param privateKeyFile Path to RSA private key file in PKCS8 format
     * @param genericAPIKeay
     * @param baseURL
     * @throws Exception
     */
    public UsersProvisioning(String publicKeyFile, String privateKeyFile, String genericAPIKeay, String baseURL) throws Exception {
        this(RSAUtils.loadPublicKey(publicKeyFile), RSAUtils.loadPrivateKey(privateKeyFile), genericAPIKeay, baseURL);
    }

    /**
     * Checks if user exists
     *
     * @param email User email
     * @return True if user exists
     */
    public boolean isUserExists(String email) {
        Map<String, Object> response = sendRequestAndParse(email, IS_USER_EXISTS_URL);
        LOG.debug("useExists resp=" + response);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Adds user
     *
     * @param user User
     * @return True if user has been successfully added
     */
    public boolean addUser(User user) {
        Map<String, Object> response = sendRequestAndParse(user, ADD_USER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Edits user
     *
     * @param user User
     * @return True if user has been successfully edited
     */
    public boolean editUser(User user) {
        Map<String, Object> response = sendRequestAndParse(user, EDIT_USER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Suspends an user
     *
     * @param email User email
     * @return True if user has been successfully suspended
     */
    public boolean suspendUser(String email) {
        Map<String, Object> response = sendRequestAndParse(email, SUSPEND_USER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Deletes a user
     *
     * @param email User email
     * @return True if user has been successfully deleted
     */
    public boolean deleteUser(String email) {
        Map<String, Object> response = sendRequestAndParse(email, DELETE_USER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Resends activation email
     *
     * @param email User email
     * @return True if user has been successfully sent
     */
    public boolean resendActivationEmail(String email) {
        Map<String, Object> response = sendRequestAndParse(email, RESEND_ACTIVATION_EMAIL_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Adds group
     *
     * @param group Group with following the attributes: { "Name": "Group name",
     * "OrganizationalUnit": "Org unit" , }
     * @return Group guid
     */
    public String addGroup(Group group) {
        return (String) sendRequestAndParse(group, ADD_GROUP_URL, String.class);
    }

    public Map<String, Object> getGroups(int size, int offSet) {
        return (Map) sendRequestAndParse(new Pagination(size, offSet), GET_GROUPS);
    }

    /**
     * Deletes a group
     *
     * @param group Group name
     * @return True if group has been successfully deleted
     */
    public boolean deleteGroup(String group) {
        Map<String, Object> response = sendRequestAndParse(group, DELETE_GROUP_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Assigns a group member
     *
     * @param groupMember Group member with the following attributes: {
     * "GroupName": "Group name", "MemberName": "user@test.com" }
     * @return True if group member has been successfully added
     */
    public boolean assignGroupMember(Map groupMember) {
        Map<String, Object> response = sendRequestAndParse(groupMember, ASSIGN_GROUP_MEMBER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Deletes a group member
     *
     * @param groupMember Group member with the following attributes: {
     * "GroupName": "Group name", "MemberName": "user@test.com", }
     * @return True if group member has been successfully deleted
     */
    public boolean unassignGroup(Map groupMember) {
        Map<String, Object> response = sendRequestAndParse(groupMember, UNASSIGN_GROUP_MEMBER_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Updates or creates a role
     *
     * @param role Role with the following attributes: {"Id": "...", "Name":
     * "..."}
     * @return Role guid
     */
    public String postRole(Map role) {
        return sendRequestAndParse(role, POST_ROLE_URL, String.class);
    }

    /**
     * Deletes a role
     *
     * @param roleGUID Role guid
     * @return True is role successfully deleted
     */
    public boolean deleteRole(String roleGUID) {
        Map<String, Object> response = sendRequestAndParse(roleGUID, DELETE_ROLE_URL);
        return (boolean) response.get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Gets assigned items to role
     *
     * @param roleGUID Role guid
     * @return A map containing users and groups: { "Users": ["user1@test.com",
     * "user2@test.com"], "Groups": ["Group 1", "Group 1"] }
     */
    public Map getAssignedToRole(String roleGUID) {
        return (Map) sendRequestAndParse(roleGUID, GET_ASSIGNED_TO_ROLE_URL);
    }

    /**
     * Assigns items to role
     *
     * @param roleItems Role items with the following attributes: { "Code":
     * "Role guid", "Users": ["user1@test.com", "user2@test.com"], "Groups":
     * ["Group 1", "Group 1"] }
     * @return True if items has been successfully added
     */
    public boolean assignToRole(Map roleItems) {
        return (boolean) sendRequestAndParse(roleItems, ASSIGN_TO_ROLE_URL).get(ProvisioningResponse.SUCCEEDED);
    }

    /**
     * Assigns items to role
     *
     * @param size Maximum number of roles to be returned, it must contain the
     * following attributes { "Size": 10, "Offset": 0 }
     * @param offSet
     * @return A map containing the following attributes: { "Total":2, "Items":
     * [ {"Id":"guid1",Name:"Role 1",Order:1}, {"Id":"guid2",Name:"Role
     * 2",Order:2} ] }
     */
    public Map getRoles(int size, int offSet) {
        return (Map) sendRequestAndParse(new Pagination(size, offSet), GET_ROLES_URL);
    }

    /**
     *
     * @param payload Data
     * @param url Request URL
     * @return A map containing result values
     */
    public Map<String, Object> sendRequestAndParse(Object payload, String url) {
        return sendRequestAndParse(payload, url, Map.class);
    }

    /**
     * Sends a REST post request to LastPass API, parses null response
     *
     * @param <T>
     * @param payload Data
     * @param url Request URL
     * @param clazz Specifies return type
     * @return Object instance of type T
     */
    public <T> T sendRequestAndParse(Object payload, String url, Class<T> clazz) {
        try {
            GenericRequest request = new GenericRequest(genericAPIKey, payload, rsaCipher);
            RestClient client = new RestClient();
            GenericResponse response = client.post(url, request, GenericResponse.class);
            return GenericResponseHandler.parseGenericResponse(response, rsaCipher, clazz);
        } catch (JsonProcessingException ex) {
            LOG.error("Creating generic request", ex);
        } catch (IOException ex) {
            LOG.error("Error parsing generic response", ex);
        }
        return null;
    }

}

File: python-provisioning/UsersProvisioning.py
import configparser
import base64
from _datetime import datetime
import json

import requests
from http import HTTPStatus

from AESCipher import AESCipher
from RSACipher import RSACipher


class UsersProvisioning(object):
    UTF_ENCODING = "utf-16le"
    APPLICATION_JSON = "application/json"

    def __init__(self, pub_key, priv_key):
        config = configparser.RawConfigParser()
        config.read('config.properties')
        self.GENERIC_API_KEY = config.get("Provisioning", "generic_api_key")
        self.BASE_URL = config.get("Provisioning", "provisioning_url")
        self.ADD_USER_URL = self.BASE_URL + "AddUser"
        self.EDIT_USER_URL = self.BASE_URL + "EditUser"
        self.DELETE_USER_URL = self.BASE_URL + "DeleteUser"
        self.SUSPEND_USER_URL = self.BASE_URL + "SuspendUser"
        self.RESEND_ACTIVATION_EMAIL_URL = self.BASE_URL + "ResendActivationEmail"
        self.PUBLIC_KEY_REGISTRATION_URL = self.BASE_URL + "PKReg"
        self.IS_USER_EXISTS_URL = self.BASE_URL + "IsUserExist"
        self.ADD_GROUP_URL = self.BASE_URL + "AddGroup"
        self.DELETE_GROUP_URL = self.BASE_URL + "DeleteGroup"
        self.ASSIGN_GROUP_MEMBER_URL = self.BASE_URL + "AssignGroupMember"
        self.UNASSIGN_GROUP_MEMBER_URL = self.BASE_URL + "UnassignGroupMember"
        self.POST_ROLE_URL = self.BASE_URL + "PostRole"
        self.GET_ROLES_URL = self.BASE_URL + "GetRoles"
        self.DELETE_ROLE_URL = self.BASE_URL + "DeleteRole"
        self.ASSIGN_TO_ROLE_URL = self.BASE_URL + "AssignToRole"
        self.GET_ASSIGNED_TO_ROLE_URL = self.BASE_URL + "GetAssignedToRole"
        self.rsaCipher = RSACipher(pub_key, priv_key, self.UTF_ENCODING)
        self.trimmed_pub_key = pub_key \
            .replace("-----BEGIN PUBLIC KEY-----", "") \
            .replace("-----END PUBLIC KEY-----", "") \
            .replace("\n", "")

    # Users

    def is_user_exists(self, email):
        """Checks if users exists
        :rtype: bool
        :param email: user email address
        :return: True if user exists
        """
        response = self.send_request_and_parse(email, self.IS_USER_EXISTS_URL)
        return response["Succeeded"]

    def add_user(self, user: dict):
        """Adds an user
        :rtype: bool
        :param user: User dictionary:
        {
             "Email" = "email@test.com",
             "SecondaryEmail" = "email2@test.com",
             "FirstName" = "First,
             "LastName" = "Last",
             "PhoneNumber" = "(555) 555-55-55",
             "EmployeeID" = "EmployeeID",
             "Department" = "Department",
             "Description" = "Description",
             "SamAccountName" = "SamAccountName",
             "OU" = "Organization unit"
             'Attributes" = [{ Name: "Attr 1", Value: "Value 1"}, { Name: "Attr 2", Value: "Value 2"}]
         }
        :return: True if user has been successfully added
        """
        response = self.send_request_and_parse(user, self.ADD_USER_URL)
        return response["Succeeded"]

    def edit_user(self, user):
        """Edits an user
        :rtype: bool
        :param user: User dictionary:
        {
             "Email" = "email@test.com",
             "SecondaryEmail" = "email2@test.com",
             "FirstName" = "First,
             "LastName" = "Last",
             "PhoneNumber" = "(555) 555-55-55",
             "EmployeeID" = "EmployeeID",
             "Department" = "Department",
             "Description" = "Description",
             "SamAccountName" = "SamAccountName",
             "OU" = "Organization unit"
             'Attributes" = [{ Name: "Attr 1", Value: "Value 1"}, { Name: "Attr 2", Value: "Value 2"}]
         }
        :return: True if user has been successfully edited
        """
        response = self.send_request_and_parse(user, self.EDIT_USER_URL)
        return response["Succeeded"]

    def delete_user(self, email):
        """Deletes an user
        :rtype: bool
        :param email: User email
        :return: True if user has been successfully deleted
        """
        response = self.send_request_and_parse(email, self.DELETE_USER_URL)
        return response["Succeeded"]

    def suspend_user(self, email):
        """Suspends an user
        :rtype: bool
        :param email: User email
        :return: True if user has been successfully suspended
        """
        response = self.send_request_and_parse(email, self.SUSPEND_USER_URL)
        return response["Succeeded"]

    def resend_activation_email(self, email):
        """Resend activation email
        :rtype: bool
        :param email: User email
        :return: True if email activation has been successfully sent
        """
        response = self.send_request_and_parse(email, self.RESEND_ACTIVATION_EMAIL_URL)
        return response["Succeeded"]

    # Groups

    def add_group(self, group):
        """Adds a group
        :rtype: dict
        :param group: Group dictionary:
        { 
             "Name": "Group name",
             "OrganizationalUnit": "Org unit" ,
         }
        :return: A dictionary containing:
         { 
             "Succeeded: True,
             "Message": "...",
             "Value": "group guid"
         }
        """
        return self.send_request_and_parse(group, self.ADD_GROUP_URL)

    def delete_group(self, group):
        """Deletes a group
        :rtype: bool
        :param group: Group name
        :return: True if group has been successfully deleted
        """
        response = self.send_request_and_parse(group, self.DELETE_GROUP_URL)
        return response["Succeeded"]

    def assign_group_member(self, group_member):
        """Adds a group member
        :param group_member: Group member dictionary:
        {
             "GroupName": "Group name",
             "MemberName": "user@test.com",
         }
        :return: True if group member has been successfully added
        """
        response = self.send_request_and_parse(group_member, self.ASSIGN_GROUP_MEMBER_URL)
        return response["Succeeded"]

    def unassign_group_member(self, group_member):
        """Deletes a group member
        :param group_member: Group member dictionary:
        { 
             "GroupName": "Group name",
             "MemberName": "user@test.com",
         }
        :return: True if group member has been successfully deleted
        """
        response = self.send_request_and_parse(group_member, self.UNASSIGN_GROUP_MEMBER_URL)
        return response["Succeeded"]

    # Roles

    def post_role(self, role):
        """Updates or creates a role
        :rtype: dict
        :param role: A dictionary: {"Id": "...", "Name": "..."}
        :return: A dictionary containing:
         { "Succeeded": True, "Message": "...", "Value": "Role guid" }
        """
        return self.send_request_and_parse(role, self.POST_ROLE_URL)

    def delete_role(self, role_guid):
        """Deletes a role
        :rtype: dict
        :param role_guid: Role guid
        :return: True is role successfully deleted
        """
        response = self.send_request_and_parse(role_guid, self.DELETE_ROLE_URL)
        return response["Succeeded"]

    def get_assigned_to_role(self, role_guid):
        """Gets assigned items to role
        :rtype: dict
        :param role_guid: Role guid
        :return: A dictionary containing:
        {
            "Succeeded": True,
            "Message": "...",
            "Value": {
                "Users": ["user1@test.com", "user2@test.com"],
                "Groups": ["Group 1", "Group 1"]
            }
        }
        """
        return self.send_request_and_parse(role_guid, self.GET_ASSIGNED_TO_ROLE_URL)

    def assign_to_role(self, role_items):
        """Assigns items to role
        :rtype: dict
        :param role_items: Role items:
        {
             "Code": "Role guid",
             "Users": ["user1@test.com", "user2@test.com"],
             "Groups": ["Group 1", "Group 1"]
        }
        :return: True if items has been successfully added
        """
        response = self.send_request_and_parse(role_items, self.ASSIGN_TO_ROLE_URL)
        return response["Succeeded"]

    def get_roles(self, size):
        """Assigns items to role
        :rtype: dict
        :param size: Size: { "Size": 10, "Offset": 0 }
        :return: A dictionary containing:
         {
             "Succeeded: true,
             "Message": "...",
             "Value": {
                "Total":2,
                "Items": [
                    {"Id":"guid1",Name:"Role 1",Order:1},
                    {"Id":"guid2",Name:"Role 2",Order:2}
                ]
             }
         }
        """
        return self.send_request_and_parse(size, self.GET_ROLES_URL)

    def send_request_and_parse(self, payload, url):
        """Sends a REST post request to LastPass API, parses null response
        :rtype: dict
        :param payload: request data in json format
        :param url: REST url
        :return: response
        """
        response = self.send_request(payload, url)

        if response is None:
            return {"Succeeded": False}

        if response["Succeeded"] is False:
            print(response["Message"])

        return response

    def send_request(self, payload, url):
        """Sends a REST post request to LastPass API
        :param payload: Data
        :param url: REST url
        :return: json
        """
        now = str(datetime.utcnow())
        now = now[:19] + "Z"
        payload = json.dumps(payload).encode()
        payload = base64.b64encode(payload).decode()
        signature = self.rsaCipher.sign(now + payload)

        request = {
            "Payload": payload,
            "Timestamp": now,
            "Signature": signature,
            "Key": self.GENERIC_API_KEY
        }
        headers = {"Content-type": self.APPLICATION_JSON}
        response = requests.post(url, json.dumps(request), headers=headers)

        if response.status_code == HTTPStatus.OK.value:
            try:
                jresp = response.json()
            except Exception as err:
                print("json expected:", err)
                print("response =", response.content)
                return None

            if jresp["Succeeded"]:
                enc_key = self.rsaCipher.decrypt(jresp["Value"]["EncKey"])
                enc_key = json.loads(enc_key)
                aes_key = base64.b64decode(enc_key["K"])
                aes_iv = base64.b64decode(enc_key["V"])
                aes_cipher = AESCipher(aes_key, aes_iv, self.UTF_ENCODING)
                payload = aes_cipher.decrypt(jresp["Value"]["Payload"])
                return json.loads(payload)
            else:
                print("error:", jresp["Message"])
                # print("request =", json.dumps(request, indent=4))
                # print("response =", response.content)
                return None

        else:
            print("error:", response.text)

        return None

    def public_key_registration(self, now):
        """Registers the public RSA key in LastPass. The key is registered just once and can't be overwritten.
        :param now: UTC current time yyyy-mm-dd hh:mi:ss
        :return: True if the key was successfully published. False otherwise.
        """
        now = now[:19] + "Z"
        signature = self.rsaCipher.sign(now + self.trimmed_pub_key)

        request = {
            "Payload": self.trimmed_pub_key,
            "Timestamp": now,
            "Signature": signature,
            "Key": self.GENERIC_API_KEY
        }
        headers = {"Content-type": self.APPLICATION_JSON}
        response = requests.post(self.PUBLIC_KEY_REGISTRATION_URL, json.dumps(request), headers=headers)
        jresp = response.json()

        if response.status_code == HTTPStatus.OK.value and jresp["Succeeded"]:
            print("key updated")
            return True
        else:
            print("key update error:", jresp["Message"])
            print(json.dumps(request))
            return False
